nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java

```diff
@@ -0,0 +1,366 @@
+
+        ijpIsSet = context.getProperty(INDEX_JSON_PATH).isSet();
+        tjpIsSet = context.getProperty(TYPE_JSON_PATH).isSet();
+
+        if (attrIsSet) {
+            idStrategy = IDStrategy.Attribute;
+        } else if (jpIsSet) {
+            idStrategy = IDStrategy.JSONPath;
+
+        } else {
+            idStrategy = IDStrategy.Generate;
+        }
+
+        batchSize = context.getProperty(BATCH_SIZE).evaluateAttributeExpressions().asInteger();
+        clientService = context.getProperty(CLIENT_SERVICE).asControllerService(ElasticSearchClientService.class);
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        List<FlowFile> flowFiles = session.get(batchSize);
+        if (flowFiles.size() == 0) {
+            return;
+        }
+
+        try {
+            List<FlowFileJsonDescription> jsonStrings = getJsonStringsFromFlowFiles(flowFiles, context, session);
+            List<IndexOperationRequest> operations = buildIndexOperations(jsonStrings);
+
+            clientService.bulk(operations);
+
+            session.transfer(flowFiles, REL_SUCCESS);
+        } catch (ElasticSearchError ese) {
+            String msg = String.format("Encountered a server-side problem with ElasticSearch. %s",
+                    ese.isElastic() ? "Moving to retry." : "Moving to failure");
+            getLogger().error(msg, ese);
+            Relationship rel = ese.isElastic() ? REL_RETRY : REL_FAILURE;
+            session.transfer(flowFiles, rel);
+        } catch (Exception ex) {
+            getLogger().error("Error during indexing...", ex);
+            session.transfer(flowFiles, REL_FAILURE);
+        }
+    }
+
+    private List<FlowFileJsonDescription> getJsonStringsFromFlowFiles(List<FlowFile> flowFiles, ProcessContext context, ProcessSession session) throws IOException, JsonProcessingError {
+        List<FlowFileJsonDescription> retVal = new ArrayList<>();
+        for (int index = 0; index < flowFiles.size(); index++) {
+            final FlowFile ff = flowFiles.get(index);
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
+            session.exportTo(ff, out);
+            out.close();
+
+            String json = new String(out.toByteArray()).trim();
+
+            if (idStrategy == IDStrategy.Attribute && json.startsWith("[")) {
+                throw new JsonProcessingError("You cannot use the attribute ID strategy on an array of JSON items.");
+            }
+
+            String targetIndex = context.getProperty(INDEX).evaluateAttributeExpressions(ff).getValue();
+            String targetType  = context.getProperty(TYPE).isSet()
+                ? context.getProperty(TYPE).evaluateAttributeExpressions(ff).getValue()
+                : null;
+
+            String id = context.getProperty(ID_ATTRIBUTE).isSet()
+                ? context.getProperty(ID_ATTRIBUTE).evaluateAttributeExpressions(flowFiles.get(index)).getValue()
```

Comment:
The name ID_ATTRIBUTE is misleading - it looks like this just gets evaluated as a literal (with expression language). I think the property and it's description should exclude mention of attributes and IDStrategy.Attribute should instead be IDStrategy.Literal

Reply:
Since it comes from a flowfile attribute, I don't agree that it's misleading. A case could be made that the ID strategy should be configured explicitly, but I don't see the problem here.

Reply:
I think just considering the context of other Elasticsearch processors, users may be misled. Take PutElasticsearchHttp for example, ID_ATTRIBUTE _always_ refers to the name of an attribute. Here, ID_ATTRIBUTE is used as a literal value supporting expression language.

https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchHttp.java#L278